nixos/displayManager: make autoLogin options independent of DM type

worldofpeace opened the pull request:
Co-authored-by: volth <volth@volth.com>

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
This code is originally from @volth and split out from https://github.com/NixOS/nixpkgs/pull/81652. @hedning @jtojnar I believe this is exactly what we talked about doing after the display-manager rewrite PR.

Not sure about changing session options yet https://github.com/NixOS/nixpkgs/pull/81652#issuecomment-595813575

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

worldofpeace committed:
nixos/displayManager: make autoLogin options independent of DM type

Co-authored-by: volth <volth@volth.com>

worldofpeace commented:
Will add 20.09 notes.

hedning commented:
> Not sure about changing session options yet #81652 (comment)?

Right, I guess it depends somewhat on the semantics of `defaultSession`. We currently have this:
>           Graphical session to pre-select in the session chooser (only effective for GDM and LightDM).
>
>           On GDM, LightDM and SDDM, it will also be used as a session for auto-login.

I think having the session bundled with the `autoLogin` makes the most sense UI wise. So one option would be to just add a `autoLogin.session` and make it take priority here:
https://github.com/hedning/nixpkgs/blob/40197cd67c41a8ec6714004dd45db11521d2cb18/nixos/modules/services/x11/display-managers/default.nix#L258-L264

We could then make `defaultSession` either just do «Graphical session to pre-select in the session chooser (only effective for GDM and LightDM)», or leave it alone. Neither option is super, since the pre-selection thing only works in GDM and LightDM, and the auto-login stuff would be sorta superfluous.

volth commented on nixos/modules/services/x11/display-managers/gdm.nix:
```diff
@@ -37,6 +37,12 @@ let
+      "Set the option `services.xserver.displayManager.autoLogin.enable' instead.")
+    (mkRemovedOptionModule [ "services" "xserver" "displayManager" "gdm" "autoLogin" "user" ]
+      "Set the option `services.xserver.displayManager.autoLogin.user' instead.")
+  ];
```
If we would include in to 20.03 (it is not too late, right?), we do not need these `mkRemovedOptionModule` because these options have been introduced few weeks ago and did not yet land to any stable release (20.03 is going to be the first)

worldofpeace replied:
20.03 is a GO https://discourse.nixos.org/t/go-no-go-meeting-nixos-20-03-markhor/6495/23?u=worldofpeace. So this change won't see that release.

volth commented on nixos/modules/services/x11/display-managers/lightdm.nix:
```diff
@@ -82,6 +82,10 @@ in
+    (mkRemovedOptionModule [ "services" "xserver" "displayManager" "lightdm" "autoLogin" "enable" ]
+      "Set the option `services.xserver.displayManager.autoLogin.enable' instead.")
+    (mkRemovedOptionModule [ "services" "xserver" "displayManager" "lightdm" "autoLogin" "user" ]
+      "Set the option `services.xserver.displayManager.autoLogin.user' instead.")
```
and here

volth commented on nixos/modules/services/x11/display-managers/sddm.nix:
```diff
@@ -78,6 +78,10 @@ in
+    (mkRemovedOptionModule [ "services" "xserver" "displayManager" "sddm" "autoLogin" "enable" ]
+      "Set the option `services.xserver.displayManager.autoLogin.enable' instead.")
+    (mkRemovedOptionModule [ "services" "xserver" "displayManager" "sddm" "autoLogin" "user" ]
+      "Set the option `services.xserver.displayManager.autoLogin.user' instead.")
```
and here

worldofpeace commented on nixos/modules/services/x11/display-managers/gdm.nix:
```diff
@@ -37,6 +37,12 @@ let
 
 {
+  imports = [
+    (mkRemovedOptionModule [ "services" "xserver" "displayManager" "gdm" "autoLogin" "enable" ]
```
Yeah, I'm just now noticing that and it's a bit strange. Will do :+1:

worldofpeace commented:
Ok, I've updated the manual (before it didn't build) and added a release note.
As a result of needing to fix the manual, I had to change `autoLogin` to a submodule.

worldofpeace commented:
@GrahamcOfBorg test sddm

worldofpeace commented:
Rebased.

worldofpeace commented:
@GrahamcOfBorg test gnome3

volth commented on nixos/modules/services/x11/display-managers/default.nix:
```diff
@@ -332,12 +332,45 @@ in
+          };
+        };
+        
+        default = {};
```
Thanks for merging!

One minor optimization: `autoLogin` shouldn't be an `submodule` defaulting to `{}`, it is merely a name scope, so just
```nix
autoLogin.enable = mkOption {
}
autoLogin.user = mkOption {
}
```
would be enough.

It is how it was in https://github.com/NixOS/nixpkgs/pull/81652 which I am using for ages.

volth replied:
Remaining diff after resolving merge conflict there: https://github.com/NixOS/nixpkgs/pull/81652/files#diff-05435704a2787433f3120047d8dc2b35

worldofpeace replied:
@volth I had to opt to use the submodule because I couldn't reference it in the manual without that type.

volth replied:
Do you mean the manual generation script does not support options with `.` ?
But it is too common in nixpkgs (especially in `boot.` options), do they have problems with manual as well?
Then it is a bug to file and fix.

jtojnar replied:
It is more of addressing issue. Options have anchor assigned but arbitrary attributes on the option path do not. I usually address a group by linking to its `enable` option when available.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/autologin-appears-in-manual-but-does-not-work/11301/2
